Make `std::os::fd` public.

sunfishcode opened the pull request:
`std::os::fd` defines types like `OwnedFd` and `RawFd` and is common
between Unix and non-Unix platforms that share a basic file-descriptor
concept. Rust currently uses this internally to simplify its own code,
but it would be useful for external users in the same way, so make it
public.

This means that `OwnedFd` etc. will all appear in three places, for
example on unix platforms:
 - `std::os::fd::OwnedFd`
 - `std::os::unix::io::OwnedFd`
 - `std::os::unix::prelude::OwnedFd`

r? @joshtriplett

sunfishcode committed:
Make `std::os::fd` public.

`std::os::fd` defines types like `OwnedFd` and `RawFd` and is common
between Unix and non-Unix platforms that share a basic file-descriptor
concept. Rust currently uses this internally to simplify its own code,
but it would be useful for external users in the same way, so make it
public.

This means that `OwnedFd` etc. will all appear in three places, for
example on unix platforms:
 - `std::os::fd::OwnedFd`
 - `std::os::unix::io::OwnedFd`
 - `std::os::unix::prelude::OwnedFd`

rust-highfive commented:
Hey! It looks like you've submitted a new PR for the library teams!

If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `@rustbot label +T-libs-api -T-libs` to tag it appropriately. If this PR contains changes to any unstable APIs please edit the PR description to add a link to the relevant [API Change Proposal](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) or [create one](https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29) if you haven't already. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.

Examples of `T-libs-api` changes:

* Stabilizing library features
* Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
* Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
* Changing public documentation in ways that create new stability guarantees
* Changing observable runtime behavior of library APIs

sunfishcode committed:
Update asrawfd.js.

dtolnay commented:
Would it be possible to get a listing of everything that is being proposed to be stabilized at a path inside `std::os::fd` (and on what platforms) by this PR? Currently the summary just says "types like OwnedFd and RawFd" which isn't specific enough for me to approve the stabilization.

@rfcbot concern what are the contents

sunfishcode commented:
@dtolnay The contents of the `fd` module in this PR are:
 - types: `OwnedFd`, `BorrowedFd`, `RawFd`
 - traits: `AsFd`, `AsRawFd`, `IntoRawFd`, `FromRawFd`

The platforms this PR adds the `fd` module to are Unix-family platforms (`cfg(unix)`) and WASI (`cfg(target_os = "wasi")`).

In the future, it's possible that fortanix_sgx and/or solid may want to add this module as well, as they also have similar types and traits.

dtolnay commented:
Great!

@rfcbot resolve what are the contents
@rfcbot reviewed

BurntSushi commented:
Is this insta-stable?

The other thing I'm confused about here is why things like `RawFd` are going in a platform independent module? They aren't used on Windows, for example, right?

dtolnay commented:
I think the idea is that we already have `std::os::unix`, and `std::os::wasi`, and in the future perhaps `std::os::fortanix_sgx` or `std::os::solid`, all of which contain identical fd-related wrappers and traits. If someone is writing code for "anything with an fd", they can't easily do it. So `std::os::fd` serves as `std::os::[anything with an fd]`.

BurntSushi commented:
@dtolnay Ah I see thanks, I think that makes sense.

@sunfishcode Can we add that guidance to the docs of `std::os::fd`? That is, call out that `RawFd` (and company) aren't purely platform independent, but that they are used on so many platforms that it can be convenient to just import it from `std::os::fd`.

I do admit though that this feels a little strange to me to be honest. I like the practicality behind this, but I think you still have to gate `std::os::fd` imports with conditional compilation in order for it to be correct, right? If so, I'd expect that to be in the docs too, because the import doesn't really hint that you need any conditional compilation.

sunfishcode committed:
Clarify that the `fd` module is supported on Unix and WASI.

sunfishcode commented:
> Is this insta-stable?

Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?

> @sunfishcode Can we add that guidance to the docs of `std::os::fd`? That is, call out that `RawFd` (and company) aren't purely platform independent, but that they are used on so many platforms that it can be convenient to just import it from `std::os::fd`.

I've clarified the comment to explicitly mention both Unix and WASI.

> I do admit though that this feels a little strange to me to be honest. I like the practicality behind this, but I think you still have to gate `std::os::fd` imports with conditional compilation in order for it to be correct, right? If so, I'd expect that to be in the docs too, because the import doesn't really hint that you need any conditional compilation.

It does already have conditional compilation; the `pub mod fd` is gated on `#[cfg(any(unix, target_os = "wasi", doc))]`. Is that what you're referring to?

BurntSushi commented:
> It does already have conditional compilation; the `pub mod fd` is gated on `#[cfg(any(unix, target_os = "wasi", doc))]`. Is that what you're referring to?

Hmmm, I suppose, yes. I was more referring to consumers of this API. The import path doesn't really make it clear (like other import paths do) that conditional compilation is needed. It's just a departure from what we've done, I think.

The fact that `pub mod fd` is gated on `any(unix, target_os = "wasi", doc)` does make it better. But still overall feels a little weird to me.

BurntSushi commented:
> Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?

I'm not sure of the underlying mechanism, but basically, if we don't have to do this as insta-stable, then we probably shouldn't. The typical insta-stable things are trait impls, and I believe that's because there just isn't any alternative.

sunfishcode commented:
> > Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?
> 
> I'm not sure of the underlying mechanism, but basically, if we don't have to do this as insta-stable, then we probably shouldn't. The typical insta-stable things are trait impls, and I believe that's because there just isn't any alternative.

I've now filed a tracking issue https://github.com/rust-lang/rust/issues/98699, and updated the PR to use a `#[unstable(...)]` attribute. This was a little tricky, because the `fd` module is already used internally in a way that doesn't want it to be `#[unstable(...)]`, so it appears this requires having two separate modules.

dtolnay commented:
@rfcbot fcp cancel

m-ou-se commented on library/std/src/os/fd/raw.rs:
```diff
@@ -40,10 +40,7 @@ pub trait AsRawFd {
-    /// use std::os::unix::io::{AsRawFd, RawFd};
-    /// #[cfg(target_os = "wasi")]
-    /// use std::os::wasi::io::{AsRawFd, RawFd};
+    /// use std::os::fd::{AsRawFd, RawFd};
```
This shouldn't compile, if the stability attributes work correctly.

sunfishcode replied:
It does compile, somehow. The doc tests pass. I don't understand stability attributes enough to know what's going on.

Mark-Simulacrum commented:
I believe type aliases should allow for different stability attributes (recently used by the move of CStr to core and CString to liballoc).

sunfishcode commented:
Applying the attributes to the module is what the current PR does.

I can see how to use type aliases for the types, but I don't see a way to do the same thing for traits.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7141361663?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/dc0d7c6391ae75533ed4e0c8291b6d27dc9c098e/checks/7141361663/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing alias-3.js ... OK
Testing alias-4.js ... OK
Testing alias.js ... OK
Testing asrawfd.js ... FAILED
==> Result not found in 'others': '{"path":"std::os::fd::RawFd","name":"as_raw_fd"}'
Diff of first error:
-     "path": "std::os::fd::RawFd",
+     "path": "std::os::unix::io",
+     "path": "std::os::unix::io",
     "name": "as_raw_fd",
Testing basic.js ... OK
Testing deduplication.js ... OK
Testing enum-option.js ... OK
Testing filter-crate.js ... OK
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #99849) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7962807865?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5674f76a16556f65399afaaedc160c219f077cbc/checks/7962807865/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing alias-3.js ... OK
Testing alias-4.js ... OK
Testing alias.js ... OK
Testing asrawfd.js ... FAILED
==> Result not found in 'others': '{"path":"std::os::fd::RawFd","name":"as_raw_fd"}'
Diff of first error:
-     "path": "std::os::fd::RawFd",
+     "path": "std::os::unix::io",
+     "path": "std::os::unix::io",
     "name": "as_raw_fd",
Testing basic.js ... OK
Testing deduplication.js ... OK
Testing enum-option.js ... OK
Testing filter-crate.js ... OK
```

</details>

sunfishcode committed:
Re-introduce `unstable` attributes.

Add `#[unstable(feature = "os_fd", issue = "98699")]` to the new
`pub use` declarations.

sunfishcode commented:
I've now rebased this on https://github.com/rust-lang/rust/pull/99723, which simplifies it.

I am not sure whether the stability attributes are working as expected though.

bors commented:
:umbrella: The latest upstream changes (presumably #96324) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
Since this isn't insta-stable anymore:

@bors r+

bors commented:
:pushpin: Commit 7d80510c16bd2172427aff61e5d929aa695cdfa5 has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett 7d80510c16bd2172427aff61e5d929aa695cdfa5 -->
<!-- homu: {"type":"Approved","sha":"7d80510c16bd2172427aff61e5d929aa695cdfa5","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Dylan-DPC commented:
@bors r- 

failed in [rollup](https://github.com/rust-lang/rust/pull/101282#issuecomment-1234513560)

sunfishcode committed:
Fix compilation of the doc tests on Windows.

sunfishcode commented:
Fixed the WIndows build issue.

r? @joshtriplett

joshtriplett commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a7f3ba9c135fff78728eb7790aae7907a7621922 has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett a7f3ba9c135fff78728eb7790aae7907a7621922 -->
<!-- homu: {"type":"Approved","sha":"a7f3ba9c135fff78728eb7790aae7907a7621922","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on library/std/src/os/fd/mod.rs:
```diff
@@ -1,16 +1,25 @@
+//! This module is supported on Unix platforms and WASI, which both use a
+//! similar file descriptor system for referencing OS resources.
 
 #![stable(feature = "io_safety", since = "1.63.0")]
```
This should be `1.66.0`, right?

sunfishcode replied:
Good catch; I've now filed https://github.com/rust-lang/rust/pull/106867 to fix this.

RalfJung commented on library/std/src/os/wasi/io/mod.rs:
```diff
@@ -1,12 +1,6 @@
 #![unstable(feature = "wasi_ext", issue = "71213")]
 
-mod fd;
-mod raw;
```
This removed two `mod` statements but the corresponding files still exist. I think they are dead now? Or is there some magic going on that somehow loads them anyway?

sunfishcode replied:
They're indeed dead now. I've now filed #135512 to  remove them.

sunfishcode replied:
And I've now noticed that https://github.com/rust-lang/rust/pull/135491 already does this, so closing in favor off that.
